File: algo/src/main/java/com/hartwig/actin/algo/evaluation/reproduction/ReproductionRuleMapper.java
package com.hartwig.actin.algo.evaluation.reproduction;

import java.util.Map;

import com.google.common.collect.Maps;
import com.hartwig.actin.algo.evaluation.FunctionCreator;
import com.hartwig.actin.algo.evaluation.RuleMapper;
import com.hartwig.actin.algo.evaluation.RuleMappingResources;
import com.hartwig.actin.treatment.datamodel.EligibilityRule;

import org.jetbrains.annotations.NotNull;

public class ReproductionRuleMapper extends RuleMapper {

    public ReproductionRuleMapper(@NotNull final RuleMappingResources resources) {
        super(resources);
    }

    @NotNull
    public Map<EligibilityRule, FunctionCreator> createMappings() {
        Map<EligibilityRule, FunctionCreator> map = Maps.newHashMap();

        map.put(EligibilityRule.IS_BREASTFEEDING, isBreastfeedingCreator());
        map.put(EligibilityRule.IS_PREGNANT, isPregnantCreator());
        map.put(EligibilityRule.USES_ADEQUATE_ANTICONCEPTION, canUseAdequateAnticonceptionCreator());
        map.put(EligibilityRule.ADHERES_TO_SPERM_OR_EGG_DONATION_PRESCRIPTIONS, willingToAdhereToDonationPrescriptionsCreator());

        return map;
    }

    @NotNull
    private FunctionCreator isBreastfeedingCreator() {
        return function -> new IsBreastfeeding();
    }

    @NotNull
    private FunctionCreator isPregnantCreator() {
        return function -> new IsPregnant();
    }

    @NotNull
    private FunctionCreator canUseAdequateAnticonceptionCreator() {
        return function -> new CanUseAdequateAnticonception();
    }

    @NotNull
    private FunctionCreator willingToAdhereToDonationPrescriptionsCreator() {
        return function -> new WillingToAdhereToDonationPrescriptions();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/reproduction/ReproductionRuleMapper.kt
package com.hartwig.actin.algo.evaluation.reproduction

import com.hartwig.actin.algo.evaluation.FunctionCreator
import com.hartwig.actin.algo.evaluation.RuleMapper
import com.hartwig.actin.algo.evaluation.RuleMappingResources
import com.hartwig.actin.trial.input.EligibilityRule

class ReproductionRuleMapper(resources: RuleMappingResources) : RuleMapper(resources) {

    override fun createMappings(): Map<EligibilityRule, FunctionCreator> {
        return mapOf(
            EligibilityRule.IS_BREASTFEEDING to isBreastfeedingCreator(),
            EligibilityRule.IS_PREGNANT to isPregnantCreator(),
            EligibilityRule.USES_ADEQUATE_ANTICONCEPTION to canUseAdequateAnticonceptionCreator(),
            EligibilityRule.ADHERES_TO_SPERM_OR_EGG_DONATION_PRESCRIPTIONS to willingToAdhereToDonationPrescriptionsCreator()
        )
    }

    private fun isBreastfeedingCreator(): FunctionCreator {
        return { IsBreastfeeding(evaluationLabels.reproduction) }
    }

    private fun isPregnantCreator(): FunctionCreator {
        return { IsPregnant(evaluationLabels.reproduction) }
    }

    private fun canUseAdequateAnticonceptionCreator(): FunctionCreator {
        return { CanUseAdequateAnticonception(evaluationLabels.reproduction) }
    }

    private fun willingToAdhereToDonationPrescriptionsCreator(): FunctionCreator {
        return { WillingToAdhereToDonationPrescriptions(evaluationLabels.reproduction) }
    }
}